third_party/move/move-vm/runtime/src/execution_tracing/loggers.rs

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! Interfaces and implementations for [Trace] collection.
+
+use crate::{
+    execution_tracing::{
+        trace::{CondBrTrace, DynamicCall},
+        Trace,
+    },
+    LoadedFunction,
+};
+use move_core_types::function::ClosureMask;
+
+/// Interface for recording the trace at runtime. It is sufficient to record branch decisions as
+/// well as dynamic function calls originating from closures.
+pub trait TraceLogger {
+    /// Returns true if the trace is being collected.
+    fn is_enabled(&self) -> bool;
+
+    /// Called in the end of execution to produce a final trace, suitable for replay.
+    fn finish(self) -> Trace;
+
+    /// Called after successful execution of a bytecode instruction. It is crucial that the trace
+    /// records only successful instructions.
+    fn record_successful_instruction(&mut self);
+
+    /// Called for successful every taken conditional branch.
+    fn record_branch_taken(&mut self);
+
+    /// Called for every successful non-taken conditional branch.
```

Comment:
Why not have one function with `bool` param?

Reply:
In case rust does not optimize bool param, but probably overkill - added param, I can check asm later to see if this is inlined and optimized as expected